삭제한 채널이 구독 목록에 남아있는 현상

# 버그 리포트

## 어떤 버그인가요?

> 버그에 대해 알려주세요

삭제한 채널이 구독 목록에 남아있는 현상

<br><br>

## 어떤 상황에서 겪으셨나요?

> 버그를 겪으신 상황을 알려주세요

채널 구독이 제대로 삭제되지 않아서 FK 문제로 채널이 삭제되지 않고 있습니다

<br><br>

## 참고할만한 자료가 있을까요?

> 참고자료가 있다면 첨부해주세요

![image](https://user-images.githubusercontent.com/53105735/196844005-21b41ea9-4482-4d6d-86ee-e8ae609b03d6.png)

에러 로그 일부 (id같은 값들은 일부러 XXX로 변경했습니다)

```
[INFO ] 2022-10-20 11:40:33 [http-nio-8080-exec-10] [LoggingAspect:?] - [Request] Controller: SlackEventController / Method: save / Arguments: requestBody={"token":"XXX","team_id":"XXX","api_app_id":"XXX","event":{"channel":"XXX","actor_id":"XXX","type":"channel_deleted","event_ts":"1666233269.019300"},"type":"event_callback","event_id":"XXX","event_time":1666233269,"authorizations":[{"enterprise_id":null,"team_id":"XXX","user_id":"XXX","is_bot":true,"is_enterprise_install":false}],"is_ext_shared_channel":false}
[WARN ] 2022-10-20 11:40:33 [http-nio-8080-exec-10] [SqlExceptionHelper:?] - SQL Error: 1451, SQLState: 23000
[ERROR] 2022-10-20 11:40:33 [http-nio-8080-exec-10] [SqlExceptionHelper:?] - Cannot delete or update a parent row: a foreign key constraint fails (`dev`.`channel_subscription`, CONSTRAINT `XXX` FOREIGN KEY (`channel_id`) REFERENCES `channel` (`id`))
[INFO ] 2022-10-20 11:40:33 [http-nio-8080-exec-10] [AbstractBatchImpl:?] - HHH000010: On release of batch it still contained JDBC statements
[ERROR] 2022-10-20 11:40:33 [http-nio-8080-exec-10] [ControllerAdvice:?] - 예상하지 못한 에러가 발생하였습니다.
org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [null]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
```

<br><br>

--- a/backend/src/main/java/com/pickpick/channel/domain/ChannelSubscriptionRepository.java
+++ b/backend/src/main/java/com/pickpick/channel/domain/ChannelSubscriptionRepository.java
@@ -25,6 +25,8 @@ public interface ChannelSubscriptionRepository extends Repository<ChannelSubscri
 
     void deleteAllByChannelAndMember(Channel channel, Member member);
 
+    void deleteAllByChannelSlackId(String channelSlackId);
+
     default ChannelSubscription getFirstByMemberIdOrderByViewOrderAsc(final Long memberId) {
         return findFirstByMemberIdOrderByViewOrderAsc(memberId)
                 .orElseThrow(() -> new SubscriptionNotFoundException(memberId));

--- a/backend/src/main/java/com/pickpick/slackevent/application/channel/ChannelDeletedService.java
+++ b/backend/src/main/java/com/pickpick/slackevent/application/channel/ChannelDeletedService.java
@@ -1,6 +1,7 @@
 package com.pickpick.slackevent.application.channel;
 
 import com.pickpick.channel.domain.ChannelRepository;
+import com.pickpick.channel.domain.ChannelSubscriptionRepository;
 import com.pickpick.message.domain.MessageRepository;
 import com.pickpick.slackevent.application.SlackEvent;
 import com.pickpick.slackevent.application.SlackEventService;
@@ -14,10 +15,14 @@ import org.springframework.stereotype.Service;
 public class ChannelDeletedService implements SlackEventService {
 
     private final ChannelRepository channels;
+    private final ChannelSubscriptionRepository channelSubscriptions;
     private final MessageRepository messages;
 
-    public ChannelDeletedService(final ChannelRepository channels, final MessageRepository messages) {
+    public ChannelDeletedService(final ChannelRepository channels,
+                                 final ChannelSubscriptionRepository channelSubscriptions,
+                                 final MessageRepository messages) {
         this.channels = channels;
+        this.channelSubscriptions = channelSubscriptions;
         this.messages = messages;
     }
 
@@ -26,6 +31,7 @@ public class ChannelDeletedService implements SlackEventService {
         String channelSlackId = extractChannelSlackId(requestBody);
 
         messages.deleteAllByChannelSlackId(channelSlackId);
+        channelSubscriptions.deleteAllByChannelSlackId(channelSlackId);
         channels.deleteBySlackId(channelSlackId);
     }
 

--- a/backend/src/test/java/com/pickpick/slackevent/application/channel/ChannelDeletedServiceTest.java
+++ b/backend/src/test/java/com/pickpick/slackevent/application/channel/ChannelDeletedServiceTest.java
@@ -2,6 +2,7 @@ package com.pickpick.slackevent.application.channel;
 
 import static com.pickpick.fixture.ChannelFixture.NOTICE;
 import static com.pickpick.fixture.MemberFixture.HOPE;
+import static com.pickpick.fixture.MemberFixture.KKOJAE;
 import static com.pickpick.fixture.MessageFixtures.PLAIN_20220712_18_00_00;
 import static com.pickpick.fixture.MessageFixtures.PLAIN_20220715_14_00_00;
 import static com.pickpick.fixture.WorkspaceFixture.JUPJUP;
@@ -11,6 +12,8 @@ import static org.junit.jupiter.api.Assertions.assertAll;
 
 import com.pickpick.channel.domain.Channel;
 import com.pickpick.channel.domain.ChannelRepository;
+import com.pickpick.channel.domain.ChannelSubscription;
+import com.pickpick.channel.domain.ChannelSubscriptionRepository;
 import com.pickpick.member.domain.Member;
 import com.pickpick.member.domain.MemberRepository;
 import com.pickpick.message.domain.MessageRepository;
@@ -32,6 +35,9 @@ class ChannelDeletedServiceTest {
     private ChannelRepository channels;
 
     @Autowired
+    private ChannelSubscriptionRepository channelSubscriptions;
+
+    @Autowired
     private MessageRepository messages;
 
     @Autowired
@@ -51,9 +57,25 @@ class ChannelDeletedServiceTest {
         databaseCleaner.clear();
     }
 
-    @DisplayName("채널 삭제 이벤트가 전달되면 채널과 메시지들이 삭제된다")
+    @DisplayName("채널 삭제 이벤트가 전달되면 채널이 삭제된다")
     @Test
-    void channelAndMessagesShouldBeDeletedOnChannelDeletedEvent() {
+    void channelShouldBeDeletedOnChannelDeletedEvent() {
+        // given
+        Workspace jupjup = workspaces.save(JUPJUP.create());
+        Channel notice = channels.save(NOTICE.create(jupjup));
+
+        String request = createChannelDeletedRequest(notice);
+
+        // when
+        channelDeletedService.execute(request);
+
+        //then
+        assertThat(channels.findAll()).isEmpty();
+    }
+
+    @DisplayName("채널 삭제 이벤트가 전달되면 채널과 관련된 메시지들이 삭제된다")
+    @Test
+    void messagesShouldBeDeletedOnChannelDeletedEvent() {
         // given
         Workspace jupjup = workspaces.save(JUPJUP.create());
         Member hope = members.save(HOPE.createLogin(jupjup));
@@ -67,9 +89,30 @@ class ChannelDeletedServiceTest {
         channelDeletedService.execute(request);
 
         //then
+        assertThat(messages.findAll()).isEmpty();
+    }
+
+    @DisplayName("채널 삭제 이벤트가 전달되면 채널과 관련된 채널 구독 목록이 삭제된다")
+    @Test
+    void channelSubscriptionsShouldBeDeletedOnChannelDeletedEvent() {
+        // given
+        Workspace jupjup = workspaces.save(JUPJUP.create());
+        Channel notice = channels.save(NOTICE.create(jupjup));
+
+        Member hope = members.save(HOPE.createLogin(jupjup));
+        Member kkojae = members.save(KKOJAE.createLogin(jupjup));
+        channelSubscriptions.save(new ChannelSubscription(notice, hope, 1));
+        channelSubscriptions.save(new ChannelSubscription(notice, kkojae, 1));
+
+        String request = createChannelDeletedRequest(notice);
+
+        // when
+        channelDeletedService.execute(request);
+
+        //then
         assertAll(
-                () -> assertThat(channels.findAll()).isEmpty(),
-                () -> assertThat(messages.findAll()).isEmpty()
+                () -> assertThat(channelSubscriptions.findAllByMemberId(hope.getId())).isEmpty(),
+                () -> assertThat(channelSubscriptions.findAllByMemberId(kkojae.getId())).isEmpty()
         );
     }
 

